### PR TITLE
[Perf][Ming-Omni] Add opt-in breakable Prefill CUDA Graph support

### PR DESCRIPTION
**Pull request by leihehehe**
## Motivation

Ming-Omni thinker prefill previously composed request-specific text, image,
video, and audio embeddings in a custom eager path. That path manually invoked
the inner text model and returned `can_run_cuda_graph=False`, so it bypassed
SGLang's standard prefill dispatcher and could not use the shared breakable
Prefill CUDA Graph integration introduced in #1364.

This PR adopts the shared `OmniPrefillInputs` contract for Ming-Omni. Ming
remains responsible only for composing the current request embeddings; SGLang
remains the sole owner of graph admission, bucket selection, static-buffer
population, replay, and shape-based eager fallback.

Breakable prefill remains explicitly opt-in. This PR does not enable it by
default, prescribe a production bucket ladder, or add a Ming-specific graph
runner.

## Modifications

- Prepare the complete current prefill-window embeddings in
  `MingThinkerModelRunner.before_prefill()` for both text-only and multimodal
  requests.
- Attach those embeddings through `OmniPrefillInputs` while leaving
  upstream-owned `ForwardBatch.input_embeds` unset during graph admission.
- Remove the Ming-specific manual eager forward and delegate execution to the
  standard SGLang model runner.
- Preserve the existing multimodal placeholder replacement and chunked-prefill
  consumption behavior.
- Use the shared generation batch override builder and policy validator for
  Ming thinker configuration.
- Register the prefill `input_embeds` static slot and run shared post-capture
  attestation only when the resolved backend is `breakable`.
- Preserve whether the prefill backend was explicitly selected, so an
  operator-requested graph must materialize while an auto-selected backend may
  retain SGLang's eager fallback semantics.
- Reject an explicit backend-only `breakable` selection when no positive cap,
  bucket ladder, or chunk size can produce capture shapes.
- Fail fast if upstream `input_embeds` or `replace_embeds` fields are already
  populated instead of silently skipping Ming's sidecar attachment.
- Accept the shared `omni_prefill_rids` forward kwarg required by the sidecar
  transport contract.
- Keep the query, key, and value tensors token-major before `RadixAttention`.
  Real H200 capture exposed the previous head-shaped query layout because the
  breakable backend allocates its output from the query shape.
- Add focused tests for the sidecar, graph bootstrap, policy validation, TP
  forwarding, attestation API, and attention shape contract.

## Related Issues

Related to #1357. This PR covers the Ming-Omni adopter only and does not close
the broader tracker.

### Revisions and environment

| Item | Value |
|---|---|
| Measured GPU base | `428b73225f808f0d93669140983e90839c9aea0e` |
| Measured GPU candidate | `654e7c5c5e09b3fc29b3c8f4ab9680cc7390e054` |
| Current main base | `c42137949ebe3e4033913af8cc13c91b2850a319` |
| Current PR head | `b79afc265613972d0e4a2e8ea593506d2ead03e3` |
| Model revision | `6a2e1dec07066d20f62a743ac7c34284e4a3932d` |
| GPU | 2 x NVIDIA H200, Thinker TP2 |
| Python | 3.12.3 |
| PyTorch | 2.13.0+cu130 |
| Transformers | 5.12.1 |
| SGLang | 0.5.19 |
| FlashInfer Python | 0.6.18 |
| Model dtype | BF16 |

The measured GPU candidate completed the full Ming unit directory with
**192 passed**. After rebasing onto the stage-device contract in #1745, the
current head completed **637 relevant tests with 1 skipped and 16 deselected**
against SGLang 0.5.19.

The commits after the measured candidate add fail-fast configuration/input
validation, preserve operator-selection intent, and adopt main's unified
device/GPU resolution. For the explicit B/C configurations below, they resolve
to the same values used during measurement and do not change model math, graph
buckets, replay, or sampling.

### Real-GPU correctness and mechanism checks

The GPU qualification used three arms:

- **A:** current main, retaining the pre-adopter Ming custom eager prefill.
- **B:** this PR with prefill CUDA Graph disabled, exercising standard eager.
- **C:** the same PR commit with breakable prefill CUDA Graph enabled.

The following checks passed on both TP ranks:

- all six configured token buckets `[64, 128, 256, 512, 1024, 2048]` captured;
- original graph admission observed `ForwardBatch.input_embeds is None`;
- every replay copied the current live embedding into the static slot with
  equal content hashes;
- text, image, real-audio conditioning, same-bucket audio A -> B -> A
  isolation, chunked audio prefill, and over-cap eager fallback completed;
- the chunked-audio request replayed a later chunk with a non-zero prefix;
- all over-cap batches used explained standard eager fallback;
- A and B returned the same targeted text/image and MMSU regression outputs;
- B and C returned the same targeted text/image/audio correctness outputs.

The first real C startup also exposed and led to the token-major attention fix
in this head. After the fix, a separate real-H200 smoke run captured all six
buckets and replayed both text and audio requests with equal live/static
embedding hashes.

One pure-text case had a shape-dependent BF16 difference when 38 live tokens
were padded to the 64-token bucket. An exact 38-token graph reproduced B with
zero error on both ranks. For the padded case:

| Metric | Result |
|---|---:|
| Top-1 token | Equal |
| Cosine similarity | 0.999804 |
| Relative L2 | 1.95% |
| Maximum absolute logit difference | 0.48828125 |

The gate accepted this only after the exact-shape proof, with pinned bounds of
cosine >= 0.999, relative L2 <= 3%, and equal argmax. All other targeted
image/audio/padded/fallback comparisons passed the original elementwise
allclose check. This is consistent with the shape-dependent BF16 behavior
reported by other breakable-prefill adopters; it is not treated as bit-exact
padded execution.

## Benchmark

The formal benchmark uses the repository's unchanged
`benchmarks.eval.benchmark_omni_mmsu.run()` and `BenchmarkRunner` timing and
metrics. Every request includes pinned real MMSU audio plus its multiple-choice
question.

Protocol:

- three fresh-server counterbalanced rounds: C -> B, B -> C, C -> B;
- concurrency `1, 8, 16, 32`;
- request flags `modalities=text` and `modalities=text+audio`;
- 96 fixed distinct MMSU IDs per cell;
- one excluded warmup cohort per cell;
- **4,608 measured requests, 4,608 successful, 0 failed**.

The deployed Ming text pipeline returned text only. The `text+audio` rows
therefore measure compatibility with that requested output flag; all such rows
record `NOT_PROVIDED_BY_TEXT_PIPELINE`. They do not establish generated-audio
fulfillment or speech quality.

### Performance results

Values are medians of the three paired fresh-server rounds. Change is C versus
B using the repository's official metrics.

| Requested flag | Concurrency | B QPS | C QPS | Throughput | Mean latency |
|---|---:|---:|---:|---:|---:|
| text | 1 | 3.071 | 4.131 | **+36.5%** | **-26.9%** |
| text | 8 | 6.860 | 8.459 | **+16.9%** | **-16.2%** |
| text | 16 | 7.454 | 8.642 | **+8.4%** | **-10.0%** |
| text | 32 | 8.053 | 8.762 | **+9.1%** | **-7.7%** |
| text+audio | 1 | 3.798 | 4.688 | **+25.2%** | **-20.1%** |
| text+audio | 8 | 7.351 | 8.522 | **+13.4%** | **-13.6%** |
| text+audio | 16 | 7.631 | 9.044 | **+15.1%** | **-15.8%** |
| text+audio | 32 | 8.352 | 8.988 | **+7.2%** | **-6.0%** |

All eight median workload comparisons improved. Across the 24 individual
round/workload pairs, 21 improved and 3 regressed; the observed throughput
range was `0.870x` to `1.407x`. This does not support a claim that every
individual run is faster.

Fresh-server ready time was 75-78 seconds for B and 78-81 seconds for C.

### Replay and fallback coverage

Counting one TP rank to avoid double-counting mirrored execution:

| Route | Original prefill batches |
|---|---:|
| Graph replay | 1,320 |
| Explained standard eager fallback | 30 |
| Unknown fallback | 0 |
| Graph batch coverage | **97.78%** |

Of the 30 eager batches, 29 exceeded the 2,048-token capture cap and one
28-token batch hit the shared 2x padding guard. All 24 C performance cells
observed real replay and permitted graph-speedup attribution.

### Bounded quality result

The same 96 IDs were intentionally repeated across rounds and request flags;
this is not 4,608 independent quality samples.

| Arm | Correct | Accuracy |
|---|---:|---:|
| B | 57 / 96 | 59.38% |
| C | 56 / 96 | 58.33% |

Three IDs had deterministic B/C greedy-text differences. Two remained wrong in
both arms; one changed from correct in B to incorrect in C. The one-sample
delta repeated across cells. The bounded set therefore does **not** establish
bit-identical output or aggregate quality parity, and it is too small to make a
broad model-quality claim. A larger quality campaign would be required before
default-enabling the optimization; this PR keeps it opt-in.

A separate current-head paired quality follow-up using all 2,000 distinct MMSU
CI samples and a denser ladder is pending a fresh GPU environment. The results
above intentionally do not assume that follow-up will establish equivalence.

## Excluded current-main paths

- **Video:** excluded because current Transformers 5.x rejects Ming's existing
  Qwen2VL `videos=` preprocessing call before the Thinker runs. This is
  independent of this PR.
- **Generated speech:** excluded because the current-main Ming Talker/Audio VAE
  path has separate known failures tracked by #1114 and draft #1132.

The evidence in this PR applies to the Ming text-output pipeline and real
text/image/audio conditioning through the Thinker. It is not full multimodal or
full-model quality acceptance.

## Checklist

- [x] Format code according to the repository conventions.
- [x] Add unit tests.
- [x] Validate real breakable graph capture, replay, static-slot isolation, and
  eager fallback.
- [x] Provide three-round throughput and latency results.
- [x] Report the bounded output/accuracy difference without claiming full
  quality parity.
- [ ] For reviewers: If you have not made any contributions to this PR and are
  only assisting with merging the main branch, please remove yourself as a
  co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as long as
the label remains. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/ming_omni/bootstrap.py [hunk 1]
```diff
@@ -64,7 +71,13 @@ def create_thinker_scheduler(
         tp_rank=tp_rank,
         nccl_port=nccl_port,
         model_arch_override="BailingMoeV2ForCausalLM",
+        enable_prefill_input_embeds=enable_prefill_input_embeds,
     )
+    if enable_prefill_input_embeds:
+        attest_prefill_cuda_graphs(
+            model_worker.model_runner,
+            operator_selected=True,
```
[the remark this answers is not in the document]
Comment (leihehehe) [reply sglang_omni/models/ming_omni/bootstrap.py]: fixed

File: sglang_omni/models/ming_omni/stages.py [hunk 2]
```diff
@@ -296,19 +296,33 @@ def create_sglang_thinker_executor_from_config(
...
-    overrides.setdefault("sampling_backend", "pytorch")
-    overrides.setdefault("trust_remote_code", False)
+    operator_selected = operator_selected_prefill_backend(server_args_overrides)
+    overrides = build_generation_batch_overrides(
+        max_running_requests=16,
+        server_args_overrides=server_args_overrides,
+        disable_cuda_graph=False,
+        sampling_backend="pytorch",
+        trust_remote_code=False,
+    )
```
[the remark this answers is not in the document]
Comment (leihehehe) [reply sglang_omni/models/ming_omni/stages.py:329]: fixed by adding fail-fast validation
